File: src/frontend/my-novel-builder/src/app/types/dtos/generate/compendium-generate-text-request.dto.ts
export interface CompendiumGenerateTextRequestDto {
  model: string;
  promptId: string;
  compendiumId: string;
  contextInfo: CompendiumTextGenerationContextInfoDto;
}

export interface CompendiumTextGenerationContextInfoDto {
  $type: CompendiumTextGenerationType;
}

export interface DescribeImageContextInfoDto
  extends CompendiumTextGenerationContextInfoDto {
  $type: CompendiumTextGenerationType.DescribeImage;
  instructions: string | null;
}

export interface CreateCompendiumRecordImageGenerationPromptContextInfoDto
  extends CompendiumTextGenerationContextInfoDto {
  $type: CompendiumTextGenerationType.CreateCompendiumRecordImageGenerationPrompt;
  compendiumRecordId: string;
  instructions: string | null;
}

export enum CompendiumTextGenerationType {
  DescribeImage = 'describeImage',
  CreateCompendiumRecordImageGenerationPrompt = 'createCompendiumRecordImageGenerationPrompt',
}

File: src/backend/MyNovelBuilder/MyNovelBuilder.WebApi/Dtos/Generate/CompendiumGenerateTextRequestDto.cs
namespace MyNovelBuilder.WebApi.Dtos.Generate;

/// <summary>
/// DTO for a compendium-scoped request to generate text.
/// </summary>
public class CompendiumGenerateTextRequestDto
{
    /// <summary>
    /// The model to use for text generation.
    /// </summary>
    public required string Model { get; set; }
    
    /// <summary>
    /// The prompt ID.
    /// </summary>
    public Guid PromptId { get; set; }
    
    /// <summary>
    /// The compendium ID.
    /// </summary>
    public Guid CompendiumId { get; set; }
    
    /// <summary>
    /// The context information.
    /// </summary>
    public required CompendiumTextGenerationContextInfoDto ContextInfo { get; set; }
}
